src/Cli/dotnet/commands/dotnet-publish/Program.cs

```diff
@@ -43,30 +64,64 @@ public static PublishCommand FromParseResult(ParseResult parseResult, string msb
                 noRestore,
-                msbuildPath);
+                msbuildPath
+            );
+
+            List<string> CreatePropertyListForPublishInvocation(ParseResult parseResult) {
+                var msbuildArgs = new List<string>()
+                    {
+                        "-target:Publish",
+                        "--property:_IsPublishing=true" // This property will not hold true for MSBuild /t:Publish. VS should also inject this property when publishing in the future.
+                    };
+
+                IEnumerable<string> slnOrProjectArgs = parseResult.GetValueForArgument(PublishCommandParser.SlnOrProjectArgument);
+
+                CommonOptions.ValidateSelfContainedOptions(parseResult.HasOption(PublishCommandParser.SelfContainedOption),
+                    parseResult.HasOption(PublishCommandParser.NoSelfContainedOption));
+
+                msbuildArgs.AddRange(parseResult.OptionValuesToBeForwarded(PublishCommandParser.GetCommand()));
+                ReleasePropertyProjectLocator projectLocator = new ReleasePropertyProjectLocator(Environment.GetEnvironmentVariable(EnvironmentVariableNames.ENABLE_PUBLISH_RELEASE_FOR_SOLUTIONS) != null);
+                msbuildArgs.AddRange(projectLocator.GetCustomDefaultConfigurationValueIfSpecified(parseResult, MSBuildPropertyNames.PUBLISH_RELEASE, slnOrProjectArgs, PublishCommandParser.ConfigurationOption) ?? Array.Empty<string>());
+                msbuildArgs.AddRange(slnOrProjectArgs ?? Array.Empty<string>());
+
+                return msbuildArgs;
+            }
+
+            List<string> DiscoverPropertiesFromPublishProfile(ParseResult parseResult)
+            {
+                ReleasePropertyProjectLocator projectLocator = new ReleasePropertyProjectLocator(Environment.GetEnvironmentVariable(EnvironmentVariableNames.ENABLE_PUBLISH_RELEASE_FOR_SOLUTIONS) != null);
+                var cliProps = projectLocator.GetGlobalPropertiesFromUserArgs(parseResult);
+                var projectInstance = projectLocator.GetTargetedProject(parseResult.GetValueForArgument(PublishCommandParser.SlnOrProjectArgument), cliProps, MSBuildPropertyNames.PUBLISH_RELEASE);
+                // this can happen if the project wasn't loadable, or if we're at a solution context
+                if (projectInstance == null)
+                {
+                    return new List<string>();
+                }
+                var importedPropValue = projectInstance.GetPropertyValue(MSBuildPropertyNames.PUBLISH_PROFILE_IMPORTED);
+                if (!String.IsNullOrEmpty(importedPropValue) && bool.TryParse(importedPropValue, out var wasImported) && wasImported) {
+                    try {
+                        if (projectInstance.GetPropertyValue(MSBuildPropertyNames.PUBLISH_PROFILE_FULL_PATH) is {} fullPathPropValue) {
+                            var properties = new ProjectInstance(fullPathPropValue).ToProjectRootElement().PropertyGroups.First().Properties;
```

Comment:
While I haven't done the side-by-side comparison, I imagine that it would be faster to loop through the properties in the propertiesToForwardFromProfile, instead of loading in every property on the project into memory